auth/credentials/impersonate/impersonate.go

```diff
@@ -125,16 +130,43 @@ func NewCredentials(opts *CredentialsOptions) (*auth.Credentials, error) {
 		}
 	}
 
-	var udp auth.CredentialsPropertyProvider
-	if creds != nil {
-		udp = auth.CredentialsPropertyFunc(creds.UniverseDomain)
-	}
 	return auth.NewCredentials(&auth.CredentialsOptions{
 		TokenProvider:          auth.NewCachedTokenProvider(its, tpo),
-		UniverseDomainProvider: udp,
+		UniverseDomainProvider: universeDomainProvider,
 	}), nil
 }
 
+// resolveUniverseDomainProvider returns the default service domain for a given Cloud
+// universe, with the following precedence:
+//
+// 1. A non-empty CredentialsOptions.UniverseDomain.
+// 2. If non-nil creds, then creds.UniverseDomain.
+// 3. The default value "googleapis.com".
+//
+// This is the universe domain configured for the credentials, which will be
+// used in endpoint(s), and compared to the universe domain that is separately
+// configured for the client.
+func resolveUniverseDomainProvider(opts *CredentialsOptions, creds *auth.Credentials) auth.CredentialsPropertyProvider {
```

Comment:
This seems like common logic that could be broadened and used consistently all over, similar to what we do for quotaproject. Could we do something similar here: https://github.com/googleapis/google-cloud-go/blob/main/auth/internal/internal.go#L100

Reply:
If we move as-is this to internal, it will import `CredentialsOptions` and `Credentials`. Maybe we can wait until we see this logic duplicated elsewhere?